### DISCUSSION TITLE
Using PubSub mode of Redis, the listener is missing tracking

### DISCUSSION THREAD

**Discussion by 837895759** [open-telemetry/opentelemetry-java-instrumentation#7965] [Q&A] [answered]
Hi！

When I use publish and subscribe mode of redis, I find that the listener missing tracking.

![image](https://user-images.githubusercontent.com/26782685/222665730-633b9325-61f8-4e17-a793-8d4154c5e0e5.png)
It seems from the source code that the instrumentation of subscribe method is missing

**Comment by trask** [comment, marked as answer]
hi @837895759 can you provide some more details? or are you able to send a PR proposal to fix?

**Reply by 837895759** [reply]
hi @trask 
I am testing the publish and subscribe functions of Redis. This is part of my demo. I use Jedis or Lettuce to connect Redis. 
![image](https://user-images.githubusercontent.com/26782685/223056626-78ae87b9-4006-4f36-a8d9-6afeb40c963f.png)
![image](https://user-images.githubusercontent.com/26782685/223056771-bce70f35-c3b7-437a-a59d-7aad984825a8.png)

During the test, it was found that the publishing method could be tracked by instrumentation, but the subscription processing part of the trace was lost. This is the trace data. You can see that the trace ends after publishing, and the subscription part is missing.
![image](https://user-images.githubusercontent.com/26782685/223058283-5ebd35f5-ffdc-4b82-bd19-a25e396554d8.png)

**Reply by trask** [reply]
oh, i c, thx!

ya, I don't think we are propagating span context across jedis publish/subscribe.

do u know if jedis/redis has a concept of "headers" or somewhere that we could use to propagate span context, e.g. `traceparent`?

**Reply by 837895759** [reply]
ya, in my understanding, redis has no concept of context in publish/subscribe.

I think the span context can be spliced in the byte array of the message body for transmission, but it is not a perfect solution.

however, there are few use redis publish/subscribe.  generally, we will use message queue instead. 
but I'm curious about how you solve this problem?

**Reply by mateuszrzeszutek** [reply]
Hey @837895759 ,
In this case we'd recommend [propagating the context manually](https://opentelemetry.io/docs/instrumentation/java/manual/#context-propagation). Redis does not have a concept of message headers; and the message is always a string, which we can't modify without potentially breaking the applications that produce/consume these messages.
You could send a JSON (or other structured data) as string, and use the propagators API to inject & extract the context before/after it is de/serialized from/to string.

**Reply by 837895759** [reply]
@mateuszrzeszutek 
ok, i c, thx!
